### PR TITLE
Support `start-dev` argument when running Memory James server

### PR DESCRIPTION
**Pull request by vttranlina**
- James will auto-generate keystore file with the default setting

## Why?

- Currently, when running James server (JMAP Server),  For security reasons you are required to generate your own keystore, which you can mount into the container via a volume. -> Sometimes in the case of development, this requirement is a bit verbose

- Easier for newbies in the first step 
- Security is a low priority for memory server

## How 

- Make James support `start-dev` argument when starting (now available only for the memory server)
- If keystore file is not found & `start-dev`, James will auto-generate keystore file with the default setting

### CODE REVIEW

**Review by chibenwa — COMMENTED** [review]
Looks like a very good idea to me!

How about allowing the usage of a dev mode everywhere?

(I mean even on distributed?)

We can make start-dev the default for memory, demo and not the others...

Please leverage the demo dockerfile to rely on start-dev.

Also, please move the key generation out of JMAP-draft. Suggestion: `server/container/guice/common`

Last but not least, this is a awesome initiative. Please share about it on the mailing list: why do you want that contribution, what do you expect from it, etc...

**Comment by chibenwa** [comment]
https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1525/1/testReport/junit/org.apache.james/AuthenticatedCassandraJamesServerTest$AuthenticationFailureTest/startShouldFailOnBadPassword_GuiceJamesServer_/

org.apache.james.AuthenticatedCassandraJamesServerTest$AuthenticationFailureTest.startShouldFailOnBadPassword{GuiceJamesServer} (from org.apache.james.AuthenticatedCassandraJamesServerTest) 

```
Expecting actual:
  "com.google.inject.CreationException: Unable to create injector, see the following errors:

1) [Guice/MissingConstructor]: No injectable constructor for type RunArguments.

class RunArguments does not have a @Inject annotated constructor or a no-arg constructor.

Requested by:
1  : RunArguments.class(RunArguments.java:29)
     at SecurityKeyLoader.<init>(SecurityKeyLoader.java:82)
      \_ for 3rd parameter
     at JMAPCommonModule.configure(JMAPCommonModule.java:73)
      \_ installed by: Modules$OverrideModule -> Modules$OverrideModule -> Modules$OverrideModule -> Modules$OverrideModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> Modules$CombinedModule -> JMAPServerModule -> JMAPModule -> JMAPCommonModule

Learn more:
  https://github.com/google/guice/wiki/MISSING_CONSTRUCTOR

1 error

======================
Full classname legend:
======================
JMAPCommonModule:       "org.apache.james.jmap.draft.JMAPCommonModule"
JMAPModule:             "org.apache.james.jmap.draft.JMAPModule"
JMAPServerModule:       "org.apache.james.modules.protocols.JMAPServerModule"
Modules$CombinedModule: "com.google.inject.util.Modules$CombinedModule"
Modules$OverrideModule: "com.google.inject.util.Modules$OverrideModule"
RunArguments:           "org.apache.james.RunArguments"
SecurityKeyLoader:      "org.apache.james.jmap.draft.crypto.SecurityKeyLoader"
========================
End of classname legend:
========================
```

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]
`Also, please move the key generation out of JMAP-draft. Suggestion: server/container/guice/common` -> likely not solved

LGTM.

File: server/apps/memory-app/README.md [hunk 1]
```diff
@@ -32,6 +32,14 @@ keytool -genkey -alias james -keyalg RSA -keystore keystore
 docker run -v $PWD/keystore:/root/conf/keystore docker run apache/james:cassandra-latest
 ```
 
+In the case of quick start James without manually creating a keystore (e.g. for development), just input the command argument `start-dev` when running,
+James will auto-generate keystore file with the default setting:
+
+```
+docker run apache/james:memory-latest start-dev
```
Comment (chibenwa) [inline server/apps/memory-app/README.md:39]:
```suggestion
docker run apache/james:memory-latest --generate-keystore
```

Is more idiomatic from command line, and likely more self explanatory.

Comment (chibenwa) [reply server/apps/memory-app/README.md:39]: We should document first how to auto-generate the keystore, then explain how to set up your own...

Comment (chibenwa) [reply server/apps/memory-app/README.md:39]: Please document `--generate-keystore` in all other applications.

**Comment by vttranlina** [comment]
> We can make start-dev the default for memory, demo and not the others...

For demo, It already bash script to generate a private key., private.scr. So, we don't need generate keystore

**Comment by vttranlina** [comment]
Conflict, I will rebase to resolve that

File: server/apps/memory-app/README.md [hunk 2]
```diff
@@ -32,6 +32,15 @@ keytool -genkey -alias james -keyalg RSA -keystore keystore
 docker run -v $PWD/keystore:/root/conf/keystore apache/james:memory-latest
 ```
 
+In the case of quick start James without manually creating a keystore (e.g. for development),
+James memory server will auto-generate keystore file with the default setting that is declared in `jmap.properties` (tls.keystoreURL, tls.secret)
+Hence just run docker command:
+
+```
+docker run apache/james:memory-latest
```
Comment (chibenwa) [inline server/apps/memory-app/README.md:40]: Why not `--generate-keystore` ?

I advocate to do exactly the same that other servers...

File: server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/draft/JMAPConfigurationStartUpCheck.java [hunk 3]
```diff
@@ -69,8 +105,71 @@ private CheckResult checkSecurityKey() {
...
+            }
+        }
+    }
+
     @Override
     public String checkName() {
         return CHECK_NAME;
     }
+
+    private void generateKeystore() throws Exception {
```
Comment (chibenwa) [inline server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/draft/JMAPConfigurationStartUpCheck.java:127]: It is not the JMAP startUp checks that should generate the keystore...

**Review by chibenwa — CHANGES_REQUESTED** [review]

**Comment by chibenwa** [comment]
Did you manually tested this for JPA?

File: server/apps/memory-app/README.md (hunk 2 above)
Comment (vttranlina) [reply server/apps/memory-app/README.md:40]: Base on your suggestion: "We can make start-dev the default for memory"

File: server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/draft/JMAPConfigurationStartUpCheck.java (hunk 3 above)
Comment (vttranlina) [reply server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/draft/JMAPConfigurationStartUpCheck.java:127]: It is based on your suggestion "please move the key generation out of JMAP-draft. Suggestion: server/container/guice/common"

Can you give more detail?  
I tried to move the SecurityKeyload to `server/container/guice/common`, but see it is a bit not correct, So I only move the generateKeystore method

File: server/apps/memory-app/README.md (hunk 2 above)
Comment (chibenwa) [reply server/apps/memory-app/README.md:40]: Well, now that I see this, I really wonder why we do anything special for memory ;-)

File: server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/draft/JMAPConfigurationStartUpCheck.java (hunk 3 above)
Comment (chibenwa) [reply server/container/guice/protocols/jmap/src/main/java/org/apache/james/jmap/draft/JMAPConfigurationStartUpCheck.java:127]: I can try to get a look at it ^^

**Comment by chibenwa** [comment]
Todo:

 - [ ] Manually test all APPs

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by vttranlina** [comment]
The `ConfigurationSanitizingPerformer` looks pretty good. (y) Good job override

**Pull request merged**